Merkle tree basic features

MarcosNicolau opened the pull request:
Get the basics of the merkle tree working, which would include:

- Tree creation from a vec of string slices.
- Add new elements dynamically to the tree.

Later we'd need to cover:
- Generate a proof that if contains an element.
- Verify that a given hash is contained.

MarcosNicolau committed:
init lib structure

MarcosNicolau committed:
create basic structs

MarcosNicolau committed:
blake2 dep

MarcosNicolau committed:
changed tree structs to use Rc and RefCell

as the structs were, we would not have been able to do much given the restrictions of rust, so we need to use Rc and RefCell to check borrow rules at runtime

MarcosNicolau committed:
splitted merkle tree and conventional tree into modules

MarcosNicolau committed:
add mk create

MarcosNicolau committed:
add todo functions signature

MarcosNicolau committed:
remove Nodes type

remove Nodes type, it was confusing

MarcosNicolau committed:
add tree dynamic modifications(add, update, delete)

MarcosNicolau committed:
add gen_proof

MarcosNicolau committed:
add verify_proof

MarcosNicolau committed:
move crypto fn to utils and split it into modules

MarcosNicolau commented:
All features covered, now we need to test it! ⚠️

MarcosNicolau committed:
create mod file for utils

MarcosNicolau committed:
structure code to allow more types of mk implementations

MarcosNicolau committed:
rename full merkle tree struct

vicentevieytes commented on src/tree.rs:
```diff
@@ -0,0 +1,63 @@
+
+pub type TreeNode<T> = Rc<RefCell<Node<T>>>;
+
+impl<T> Node<T> {
```
Generic seems to add some unnecessary complexity given that a Merkle Tree would only store hashes on each node, I would just set it as type Hash.

If in the future you wanted to make the hash function mutable you could make it generic but it's not a requirement right now.

MarcosNicolau committed:
compact tree: create

vicentevieytes commented on src/tree.rs:
```diff
@@ -0,0 +1,63 @@
+pub type NodeParent<T> = Weak<RefCell<Node<T>>>;
+pub type NodeChildren<T> = Vec<Rc<RefCell<Node<T>>>>;
+pub type NodeSiblings<T> = Vec<Weak<RefCell<Node<T>>>>;
+
```
I think having the Merkle Node be the node of an n-ary tree is a bit unnecesary and it makes following along a little harder.

In the Merkle Tree constructors you end up creating a binary tree anyway and pairing the hashes by two, instead of having the childrens and siblings in a vector you could just have left_children right_children and sibling.

Even if you wanted to concatenate the hashes of the n children instead of just pairing by two, that offers no real benefit in data verification or prooving.

MarcosNicolau committed:
compact tree: get, add, delete, update leaf

vicentevieytes commented on src/utils/crypto.rs:
```diff
@@ -0,0 +1,16 @@
+    hasher.finalize().into()
+}
+
+pub fn get_hash_from_data<T: DataToHash>(el: T) -> Hash {
```
Variable "el" should have a more declarative name, "element"

MarcosNicolau committed:
compact tree: add gen_proof

MarcosNicolau committed:
compact tree: add verify_proof

vicentevieytes commented on src/mk.rs:
```diff
@@ -0,0 +1,143 @@
+                        return node;
+                    }
+                    // hash with itself
+                    [a] => {
+                        let node = Node::new(
+                            get_combined_hash(a.borrow().value, a.borrow().value),
+                            Some(vec![Rc::clone(a)]),
+                            None,
+                            None,
+                        );
+                        // does not have siblings
+                        Node::set_parent(&node, a);
+                        return node;
```
You could clone it and set it as sibling of itself, shouldn't impact performance and allows you to always be able to hash an element with it's sibling or call get_siblings without checking if there is Some(node) or None as sibling.

I don't think this is that important tho. And for the root  node you would still have to set sibling to None

vicentevieytes commented on src/mk.rs:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    fn rebuild_tree(&mut self) {
+        let tree = MerkleTree::create_tree(self.leaves.clone());
+        let root_hash = tree.borrow().value;
+        self.tree = tree;
+        self.root_hash = root_hash;
+    }
```
If you are rebuilding the tree from scratch with every modification to the data, wouldn't it be the same to just ask for the data, modify it, and create a new Merkle Tree instance with the new data?

Since you went through the trouble of making each node modifiable with RefCell and Rc, you could try to only modify the nodes that need modifying (the ones that lead from the leaf to the root node).

vicentevieytes replied:
Otherwise, it would be better to just have the tree be inmutable.

MarcosNicolau committed:
fix single node siblings

fix #3

vicentevieytes commented on src/utils/num.rs:
```diff
@@ -0,0 +1,20 @@
+// Generic function that works for any type that implements IsEven
+pub fn is_even<T: IsEven>(num: T) -> bool {
+    num.is_even()
+}
```
This feels really convoluted, but i guess it works alright.

You could just use % 2 == 0 in place and it would be really understandable.

vicentevieytes reviewed (commented):
Logic seems good to me.

Proofs with the siblings are really compact and nice.

You could trim some unnecesary abstractions and generics, i left some comments, it's up to you if you want to keep up with that structure though.

You should add comments and documentation to the public functions.

MarcosNicolau commented:
Great, here are some general answers:

- Yes, this version of the tree looks really abstract when it is really not needed. I did it just to practice and understand how smart pointers really work in rust and how much effort you'd need to abstract your code.
- That is why I built a simpler option (you can see it in the #6), this one I'll leave it like this.
-  I will try to implement rebuilding the tree by mutating the necessary nodes instead of recreating the whole tree.
- For sibling of itself, that is a current issue(see #3), which I have already fixed.
- Finally, I'll make sure to add some documentation for the final version.

vicentevieytes commented:
Great :)

MarcosNicolau committed:
Merge pull request #6 from MarcosNicolau/compact-tree

Compact tree

MarcosNicolau committed:
Merge pull request #7 from MarcosNicolau/fix/full-ml-siblings

fix single node siblings

MarcosNicolau commented:
I've just merged the compact tree version and the fix!

MarcosNicolau committed:
general refactor full merkle tree

MarcosNicolau committed:
tests for full merkle tree

MarcosNicolau committed:
compact_merkle_tree: fix get_proof err

the gen_proof was not working for even trees coz we were returning a out of bounds err instead of hashing with himself

MarcosNicolau committed:
test for compact merkle tree

MarcosNicolau committed:
Merge pull request #9 from MarcosNicolau/tests

Tests

vicentevieytes commented on src/mk/compact.rs:
```diff
@@ -0,0 +1,290 @@
+        if data.is_empty() {
+            return None;
+        }
+        let leaves: Vec<Node<[u8; 64]>> = Self::get_leaves_from(data);
+        let root_hash = Self::calculate_root(leaves.clone());
+        Some(Self { leaves, root_hash })
+    }
+
+    fn calculate_root(mut leaves: Vec<MKNode>) -> Hash {
+        while leaves.len() > 1 {
+            leaves = Self::get_parent_nodes(&leaves);
```
Could leaves be borrowed by calculate_root? so you don't have to clone on the create() function.
```rust
fn calculate_root(leaves: &Vec<MKNode>) -> Hash
```

MarcosNicolau replied:
no you can't 😢, because you need to mutate it.

MarcosNicolau committed:
made structs fields pub

MarcosNicolau committed:
move tests into its own files

MarcosNicolau committed:
implement full merkle tree contains hash

MarcosNicolau committed:
implement compact merkle tree contains hash

MarcosNicolau committed:
Merge pull request #10 from MarcosNicolau/contains-hash

Add contains hash

vicentevieytes commented:
Looks good!

MarcosNicolau committed:
full merkle tree refactor to accept arr slices

fixes #4

MarcosNicolau committed:
add get_leaf_by_idx and get_leaf_by_hash

added for both types of trees. resolves #5

MarcosNicolau committed:
docs readme instructions

MarcosNicolau committed:
Merge pull request #11 from MarcosNicolau/docs

docs readme instructions

MarcosNicolau commented:
@vicentevieytes, if it looks good to you, I'll merge it!

MarcosNicolau committed:
address clippy warnings

vicentevieytes reviewed (approved):
Looks great 👍
